File: easy/prime_palindrome.hs
factors  :: Int -> [Int]
factors n = [x | x <- [1..n], mod n x == 0]

prime  :: Int -> Bool
prime x = factors x == [1, x]

pali  :: Int -> Bool
pali x | x < 10                             = True
       | x < 100 && (div x 10 == mod x 10)  = True
       | x > 100 && (div x 100 == mod x 10) = True
       | otherwise                          = False

main :: IO ()
main = print $ last [x | x <- 2 : [3, 5..1000], pali x, prime x]

File: easy/prime_palindrome.c
#include <stdbool.h>
#include <stdio.h>

static bool prime(int a, int *p) {
	while (*p * *p <= a) {
		if (a % *p == 0)
			return false;
		p++;
	}
	return true;
}

static bool pali(int x) {
	if ((x < 10) ||
	    (x < 100 && (x / 10 == x % 10)) ||
	    (x > 100 && (x / 100 == x % 10))) {
		return true;
	}
	return false;
}

int main(void) {
	int primes[256] = { 2, 3 };
	int i, m = 3, c = 2;
	for (i = 5; i < 1000; i += 2) {
		if (prime(i, primes)) {
			primes[c++] = i;
			if (pali(i))
				m = i;
		}
	}
	printf("%d\n", m);
	return 0;
}
